src/main/scala/analysis/GlobalRegionAnalysis.scala

```diff
@@ -148,48 +106,60 @@ trait GlobalRegionAnalysis(val program: Program,
    * @return Set[DataRegion]
    */
   def checkIfDefined(dataRegions: Set[DataRegion], n: CFGPosition): Set[DataRegion] = {
-    dataRegions.map { i =>
+    var converted: Set[DataRegion] = Set.empty
+    dataRegions.foreach { i =>
       val (f, p) = mmm.findDataObjectWithSize(i.start, i.size)
       val accesses = f.union(p)
       if (accesses.isEmpty) {
         i
       } else if (accesses.size == 1) {
-        dataMap(i.start) = DataRegion(i.regionIdentifier, i.start, i.size.max(accesses.head.size))
-        dataMap(i.start)
+        if (f.contains(accesses.head)) {
+          // full access
+          dataPoolMaster(i.start, i.size, false)
+          converted = converted ++ Set(dataMap(i.start))
+        } else {
+          // partial access (we cannot determine the size)
+          dataPoolMaster(i.start, i.size, false)
+          converted = converted ++ Set(dataMap(i.start))
+        }
       } else {
-        val highestRegion = accesses.maxBy(_.start)
-        dataMap(i.start) = DataRegion(i.regionIdentifier, i.start, i.size.max(highestRegion.end - i.start))
-        dataMap(i.start)
+  //        val highestRegion = accesses.maxBy(_.start)
+  //        dataMap(i.start) = DataRegion(i.regionIdentifier, i.start, i.size.max(highestRegion.end - i.start))
+  //        dataMap(i.start)
+        dataMap.remove(i.start)
+        accesses.foreach(a => dataPoolMaster(a.start, a.size, false))
+        converted = converted ++ accesses.map(a => dataMap(a.start))
       }
     }
+    converted
   }
 
   /** Transfer function for state lattice elements.
    */
-  def localTransfer(n: CFGPosition, s: Set[DataRegion]): Set[DataRegion] = {
+  def transfer(n: CFGPosition, s: Set[DataRegion]): Set[DataRegion] = {
     n match {
       case store: MemoryStore =>
-        checkIfDefined(evalMemLoadToGlobal(store.index, store.size, store), n)
+        checkIfDefined(tryCoerceIntoData(store.index, store, store.size), n)
       case load: MemoryLoad =>
-        checkIfDefined(evalMemLoadToGlobal(load.index, load.size, load, loadOp = true), n)
+        val regions: Set[DataRegion] = tryCoerceIntoData(load.index, load, load.size)
+        checkIfDefined(regions, n)
       case assign: LocalAssign =>
         // this is a constant but we need to check if it is a data region
-        checkIfDefined(evalMemLoadToGlobal(assign.rhs, 1, assign), n)
+        checkIfDefined(tryCoerceIntoData(assign.rhs, assign, 1, noLoad = true), n)
```

Comment:
The current issues seem to be in significant part due to this line. Creating regions with a size of 1 here creates a lot of regions with incorrect sizes that cause problems.